front: match PathStep and SuggestedOP with pathStepId

emersion opened the pull request:
Last episode: https://github.com/OpenRailAssociation/osrd/pull/9967

Matching via uic/trigram/ch/etc is unreliable because:

- Two OPs can have the same trigram/ch.
- The same OP could appear multiple times at different spots in a
  path.

Moreover, we had multiple variants of the matching with slight
differences. Some of the matching logic got adjusted to account
for specific bugs, while others were left alone.

Each path step can be identified with a unique ID stored in
PathStep.id. Use that instead.

(The refactoring has been split into small digestible commits to ease bisecting and make it more obvious why type and test adjustments are necessary.)

SharglutDev commented:
Fyi : https://github.com/OpenRailAssociation/osrd/pull/10334#discussion_r1920020968

emersion committed:
front: use pathStepId in ModalSuggestedVias

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: use pathStepId in usePathfinding()

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: use pathStepId in upsertPathStep()

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: use pathStepId in formatSuggestedViasToRowVias()

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: use pathStepId in TimesStopsInput

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: drop pathStepMatchesOp() and isVia()

These have all been replaced with pathStepId checks.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: use pathStepId in ItineraryMarkers

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: use pathStepId in formatOperationalPointWithTimes()

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: unexport matchPathStepAndOp()

Everybody uses pathStepId now.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
hack: re-export matchPathStepAndOp()

emersion committed:
front: set SuggestedOP.pathStepId in formatSuggestedOperationalPoints()

Various spots need to check whether a SuggestedOP has a matching
PathStep. Ensure the pathStepId field is always set when
appropriate.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: use pathStepId in upsertPathStepsInOPs()

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: set TimesStopsRow.pathStepId in useOutputTableData()

It's useful to know whether a row has been defined as an explicit
waypoint by the user, or if it's a suggested waypoint. Ensure the
pathStepId field is always set when appropriate.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion commented:
Closing in favor of https://github.com/OpenRailAssociation/osrd/pull/17984
